addons/crm/static/src/js/forecast/forecast_model_extension.js

```diff
@@ -0,0 +1,117 @@
+/** @odoo-module alias=crm.ForecastModelExtension **/
+import ActionModel from 'web.ActionModel';
+
+/**
+ * File intended for `web` (could be used in all kind of forecast configurations)
+ * 
+ * This file contains the logic behind a special "Forecast" filter.
+ * Any such filter should set the context key {forecast_filter: 1}
+ * Another context key must also be set for the view using this model extension:
+ * {forecast_field: "date_field_name"}, which represents the date/time field on which
+ * the forecast should be applied
+ * 
+ * The main purpose is to be able to modify the domain depending on the groupby granularity
+ * when the field is a `date/time` field and is the `forecast_field`. The domain should filter
+ * records from the period (granularity) containing "now", or those where the forecast_field
+ * is not set.
+ * example:
+ *  today:          2021-04-21
+ *  granularity:    month
+ *  field name:     date_field
+ *  -> the domain would be: ['&', ['date_field', '=', false], ['date_field', '>=', '2021-04-01']]
+ */
+class ForecastModelExtension extends ActionModel.Extension {
+
+    //---------------------------------------------------------------------
+    // Public
+    //---------------------------------------------------------------------
+
+    /**
+     * @override
+     * @returns {any}
+     */
+    get(property) {
+        switch (property) {
+            case "domain": return this.getDomain();
+        }
+    }
+
+    /**
+     * @override
+     */
+    prepareState() {
+        Object.assign(this.state, {
+            forecastField: null,  // {string} forecast_field from the context (name)
+            forecastFieldType: null,  // {string} forecast_field_type (date or datetime)
+            forecastFilter: null,  // {boolean} if the "Forecast" filter is active
+            forecastGranularity: null,  // {string} the granularity on the forecast_field groupby (year, quarter, month, week, day, hour)
+        });
+    }
+
+    /**
+     * The state needs to be recomputed each time the parent model initiates a load action.
+     * @override
+     * @returns {Promise}
+     */
+    callLoad() {
+        this._computeState();
+        return super.callLoad(...arguments);
+    }
+    
+    //---------------------------------------------------------------------
+    // Actions / Getters
+    //---------------------------------------------------------------------
+
+    /**
+     * Adds a domain constraint to only get records from the start of the period containing "now",
+     * only if the "Forecast" filter is active
+     * @returns {Array[]}
+     */
+    getDomain() {
+        if (!this.state.forecastFilter) {
+            return null;
+        }
+        var startMoment = moment().startOf(this.state.forecastGranularity);
```

Comment:
Shouldn't it be computed by the Service so that we hide that kind of computation ?

Reply:
The forecastService is completely independent from the forecastModelExtension, but it's true that the moment could be instantiated in the _computeState